Serialize file paths (#14)
audeering/audobject: docs/usage.rst

    d_yaml = d.to_yaml_s()
    print(d_yaml)

Resolve file paths
------------------

Portability is a core feature of
:mod:`audobject`.

> Now assume we have an object

Reviewer: ```suggestion
Assume we have an object
```

that takes as argument the path to a file.
When we serialize the object
we want to make sure that:

1. we store the file path relative to the YAML file
2. the path is correctly expanded when we re-instantiate the object

This can be achieved using
:class:`audobject.FilePathResolver`.

.. jupyter-execute::

    class MyObjectWithFile(audobject.Object):

        @audobject.init_decorator(
            resolvers={
                'path': audobject.FilePathResolver,  # ensure portability
            }
        )
        def __init__(
                self,
                path: str,
        ):
            self.path = path

        def read(self):  # print path and content
            print(self.path)
            with open(self.path, 'r') as fp:
                print(fp.readlines())

Here, we create a file and pass it to the object.

.. jupyter-execute::

    import os
    import audeer

    root = 'root'

    res_path = os.path.join(root, 're', 'source.txt')  # root/re/source.txt
    audeer.mkdir(os.path.dirname(res_path))
    with open(res_path, 'w') as fp:
        fp.write('You found me!')

    o = MyObjectWithFile(res_path)
    o.read()

> But when we serialize the object,

Reviewer: ```suggestion
When we serialize the object,
```

> we can see that the path is

Reviewer: ```suggestion
the path is
```

stored relative to the directory
of the YAML file.

.. jupyter-execute::

    import yaml

    yaml_path = os.path.join(root, 'yaml', 'object.yaml')  # root/yaml/object.yaml
    o.to_yaml(yaml_path)

    with open(yaml_path, 'r') as fp:
        content = yaml.load(fp, Loader=yaml.Loader)
    content

When we re-instantiate the object
the path gets expanded again.

.. jupyter-execute::

    o2 = audobject.Object.from_yaml(yaml_path)
    o2.read()

> This will also work from another location
> (note that we also move the referenced file here,
> as its relative location to the YAML file must not change).

Reviewer: ```suggestion
This will also work from another location.
Note that we have to move all referenced files as well,
as their relative location to the YAML file must not change.
```

When we serialize the object
we want to make sure that:

1. we store the file path relative to the YAML file
2. the path is correctly expanded when we re-instantiate the object

This can be achieved using
:class:`audobject.FilePathResolver`.

.. jupyter-execute::

    class MyObjectWithFile(audobject.Object):

        @audobject.init_decorator(
            resolvers={
                'path': audobject.FilePathResolver,  # ensure portability
            }
        )
        def __init__(
                self,
                path: str,
        ):
            self.path = path

Here, we instantiate the object with an absolute path.

.. jupyter-execute::

    import os
    import audeer

    path = os.path.join('re', 'source.txt')  # ./re/source.txt
    path = audeer.safe_path(path)
    o = MyObjectWithFile(path)
    o.path

But when we serialize the object,
we can see that the path is
stored relative to the directory
of the YAML file.

.. jupyter-execute::

    import yaml

    yaml_path = os.path.join('yaml', 'object.yaml')  # ./yaml/object.yaml
    o.to_yaml(yaml_path)

    with open(yaml_path, 'r') as fp:
        content = yaml.load(fp, Loader=yaml.Loader)
    content

> And when we re-instantiate the object

Reviewer: ```suggestion
When we re-instantiate the object
```

the path gets expanded again.

.. jupyter-execute::

    o2 = audobject.Object.from_yaml(yaml_path)
    o2.path

This will also work from another location.

.. jupyter-execute::

    import shutil

    yaml_path_new = os.path.join('some', 'where', 'yaml', 'object.yaml')
    audeer.mkdir(os.path.dirname(yaml_path_new))
    shutil.move(yaml_path, yaml_path_new)

    o3 = audobject.Object.from_yaml(yaml_path_new)

>     o3.path

Reviewer: This example is not working for me.

I have
```python
>>> o2.path
'/home/audeering.local/hwierstorf/git/audeering/audobject/docs/tmp/re/source.txt'
```
and
```python
>>> o3.path
'/home/audeering.local/hwierstorf/git/audeering/audobject/docs/tmp/some/where/re/source.txt'
```

As we only moved the YAML file and not the actual resource file the two paths should be the same.

Reviewer: I guess your assumption is that a user would always store the resource file relative to the path where she stores the YAML file, but why should she?

Author: Yes, when the YAML file is moved, of course the referenced files have to be moved, too, and their relative locations must not change. Otherwise it would not be portable. I guess a common use-case would be that you have a YAML file and the resource files in the same folder (or maybe a sub-folder). Then when you move the whole folder to another location (or you even upload to the cloud and download it to another machine like `audmodel` does) you can load the object and it will find the referenced files.

Author: > As we only moved the YAML file and not the actual resource file the two paths should be the same.

But how should `o3` be able to find original resource file when it is not moved? Maybe we have stored YAML on another machine or in another universe :)

Reviewer: That's true, but it's not obvious for a user. If I decide to store the YAML file at another place on my machine, I will find it quite handy if the path to the resource file would still work.

I think the problem is that the absolute path that is returned somehow indicates that an absolute path is used. So maybe we should change it to always show the relative path?

Otherwise we need at least to update the documentation and highlight that the resource file has to be stored always relative (and in your example add a stage where you also move the resource file, not only the YAML file).

Author: > If I decide to store the YAML file at another place on my machine, I will find it quite handy if the path to the resource file would still work.

The simply don't use `FilePathResolver`.

> So maybe we should change it to always show the relative path?

We can only change it to a relative path in the moment the object is stored to a YAML file.

Author: I added the following:

![image](https://user-images.githubusercontent.com/10383417/125588652-bd3fe526-e36f-4432-b435-9e571a29b8db.png)

Reviewer: OK, then let's stay with the current behavior of returning the absolute file path. It will make it also easier to work with as we don't have to call `audeer.safe_path()` first.

The changes to the documentation are good. I would maybe even go one step further.
Before calling
```python
path = os.path.join('re', 'source.txt')  # ./re/source.txt
```
you could simply create that file and then later really move the file.
The only downside is that there is no nice command to create an empty file.
I think the most common one is `open(filename, 'a').close()` which doesn't look to nice.
So maybe you create the empty resource file in the hidden code part and then later simply move the file?

Author: Actually, first I was creating a file and moving it, but then thought it made the example too complicated. So I though it's easier to understand without it. But you would argue it's important to understand the concept?

Reviewer: I think in this case it would help.

It's also challenging to understand as the resource file doesn't have to be located somewhere inside the YAML root folder, but can also stay outside as in your example:

```
yaml/object.yaml
re/source.txt
```
and after moving it has to look like
```
some/where/yaml/object.yaml
some/where/re/source.txt
```

Author: I changed to:

![image](https://user-images.githubusercontent.com/10383417/125593534-140edea4-25f0-4f18-a032-e6735d00034e.png)
![image](https://user-images.githubusercontent.com/10383417/125593580-2e79bac4-fc19-4856-9de2-9f8cfb4be3be.png)

Reviewer: Nice, but your forgot to push the changes.

I think it would be even clearer if we really replace one root dir by another. I think it should be possible when you don't create the dir before calling `shutil.move()`, e.g.:

```python
new_root = os.path.join('some', 'where')
shutil.move(root, new_root)

yaml_path_new = os.path.join(new_root, 'yaml', 'object.yaml')
```

Author: Good idea. I now changed the lower part to:

![image](https://user-images.githubusercontent.com/10383417/125610493-6e8cd06e-796c-4e38-8585-5a8ec06944bc.png)

and this time I also pushed the changes :)

Reviewer: Great
